Undo and Redo

theis999 opened the pull request:
This adds the option to Ctrl+Z to undo and Ctrl+Y to redo a few types of actions.

It is built State modifiers, so there are some unrelated commits included.

The change can be broken into 4 categories:
1. Undo and Redo has been added to the shortcut menu
2. Command stack, class has been added
3. Undo and Redo handlers has been added the can reverse the user actions
4. The clickhandlers for the commands have been modified to fit into the undo and redo handlers

### Command stack
The command stack is a custom circular buffer that can handle 32 elements. In such a way that overflow overrides old elements. However it is also possible to pop items of the stack (REDO) and further it is possible to pop_back items (UNDO) reusing the items on it.

### Not included
This only observes ADD, DELETE, CHANGE, MOVE UP, and MOVE DOWN for the steps grid.
Which means MOVE 5, import steps, colouring, no-order modifier, everything template, etc. Is not included.

theis999 committed:
Added stack container for undo & redo

theis999 committed:
Added menu items for Undo & Redo

theis999 committed:
ADD, DELETE, MOVE & CHANGE step works

theis999 committed:
Added undo

theis999 committed:
Fix to delete

MortenTobiasNielsen commented on src/cMain.cpp:
```diff
@@ -527,18 +527,35 @@ void cMain::UpdateTemplateGrid(wxGrid* grid, vector<StepParameters>& steps)
+		if (change.size() > 0)
+		{
+			stack.Push({
+				.row = *grid_steps->GetSelectedRows().begin() - 1,
```
I get this access violation error crashing the tool. It seems to be an off by 1 error where it doesn't take into account that there might not be anything in the grid:
![image](https://github.com/MortenTobiasNielsen/Factorio-TAS-Generator/assets/48355480/86d687b4-82e8-4224-8246-a97f0f20aef0)

theis999 replied:
I'll be honest i probably didn't test no-selection properly. As it is so rare to get into that state, you can use ctrl+click to get into it. 
I have fix for this issue.

MortenTobiasNielsen reviewed (changes requested):
I wanted to test out the new functionality and tried to delete 3 random rows and press undo, but for some reason it 4 rows comes back. 
![image](https://github.com/MortenTobiasNielsen/Factorio-TAS-Generator/assets/48355480/ed34512a-4ed7-44a1-b8d6-45de6fe47ead)
![image](https://github.com/MortenTobiasNielsen/Factorio-TAS-Generator/assets/48355480/8c03eb02-0c7b-48f4-b54c-c7180abdaf94)
Not really sure why that would happen, it isn’t duplicating one of the rows. I thought for a second it might have taken row 10707 for some reason, but the coordinates are not the same.

I cannot seem to find the change, but it seems like the Save (file) shortcut has been overridden:
![image](https://github.com/MortenTobiasNielsen/Factorio-TAS-Generator/assets/48355480/ef6ac7a3-efbf-4da7-9994-dd635ab3d12a)

I have this setup for testing:
![image](https://github.com/MortenTobiasNielsen/Factorio-TAS-Generator/assets/48355480/c5e9aaac-53c8-4e10-aebf-f2d62c87f06a)
I first added the build step and then added the mine steps from 1 to 10 and then I deleted the build step. When I undo and redo multiple times, I sometimes get this: 
![image](https://github.com/MortenTobiasNielsen/Factorio-TAS-Generator/assets/48355480/e6af1705-c404-483b-a0ad-8ac45cf1ce38)
Where it seems like the tool thinks that it is mining the inserter even though it has not been built. It is not consequent about which row gets it. 
![image](https://github.com/MortenTobiasNielsen/Factorio-TAS-Generator/assets/48355480/d58b2d80-25a6-4f64-9f16-94217f5a9a7b)

I changed 4 to 100 and 7 to 200 and then undid the changes, but for some reason 7 turned into 6. It did though turn 4 back to 4. 
![image](https://github.com/MortenTobiasNielsen/Factorio-TAS-Generator/assets/48355480/7a73e4a7-f175-4c45-a6a3-683c4d2b734a)

theis999 commented:
Your save shortcut can be set through the shortcut changer. Ctrl+S is just the default.

theis999 committed:
Fixed issue with adding steps without any rows selected.
Fixed issue without undo/redo auto-put

theis999 commented:
It looks like the first issue you found is related to auto-put. As the extra step matches the recipe. So i have made a fix for it.

I can't seem to replicate the issue with mining, even though your explanation is good. Perhaps it was fixed with one of the other changes?

MortenTobiasNielsen commented:
True it isn't often, but would be nice if new players were able to add their first step without the tool crashing. :D

Fair point - I am just pretty sure that I haven't changed the shortcut - is it because you have changed it on the branch and it spills over to my "config"?

Well spotted. I think you are right that it is due to the recipe. It could very well be a side effect solved by one of your changes.

theis999 committed:
Changed shortcut of save to ctrl+s

MortenTobiasNielsen commented on src/cMain.cpp:
```diff
@@ -675,102 +720,163 @@ void cMain::OnChangeStepClicked(wxCommandEvent& event)
 		return;
 	}
+	int startRow = rows.at(0);
+	auto steps = DeleteSteps(rows);
```
I can see that this is populated with data, but I have no idea why it does that, given that it doesn't seem like DeleteSteps explicitly return anything other than on line 808, which shouldn't be hit.

theis999 replied:
C++ magic apparently 🤣 I have no idea how that worked either. 
I have added a return statement at the end of DeleteSteps so it is a bit more sensible.

MortenTobiasNielsen commented on src/cMain.cpp:
```diff
@@ -2737,3 +2849,112 @@ void cMain::OnSkipClicked(wxCommandEvent& event)
+void cMain::SelectRowsInGrid(vector<tuple<int, StepParameters>> rows)
+{
+	grid_steps->ClearSelection();
+	for (auto& [r, _] : rows)
```
I would be come comfortable with for (auto& row : rows). I don't really think decomposing provides any benefit here - it might though also just be the lack of a meaningful name for "r".

theis999 replied:
Personally i don't mind the name to much when it is used so close to it's definition. But i will change r => row.

MortenTobiasNielsen commented on src/cMain.cpp:
```diff
@@ -2737,3 +2849,112 @@ void cMain::OnSkipClicked(wxCommandEvent& event)
+			break;
+	}
+}
+void cMain::OnRedoMenuSelected(wxCommandEvent& event)
```
Add a new line to separate the functions.

MortenTobiasNielsen reviewed (changes requested):
I tried to delete all steps in my file and "undo", which crashed the tool: 
![image](https://github.com/MortenTobiasNielsen/Factorio-TAS-Generator/assets/48355480/36cdb1f3-39cb-4738-a681-bfe892961406)

I unfortunately still get the error with the mine step thinking that it is mining something which hasn't been build yet. 
![image](https://github.com/MortenTobiasNielsen/Factorio-TAS-Generator/assets/48355480/6f2e5beb-3dbb-4685-baa0-03670011c892)

Steps to reproduce (I hope):

1. Add build inserter step.
2. Add each of the mining steps starting with 1. 
3. Delete the Inserter step.
4. Use undo to get the inserter step back.
5. Use undo 5-6 times to remove some of the mining steps.
6. Press redo fast multiple times until all mining steps are back.

theis999 committed:
Revert "Changed shortcut of save to ctrl+s"

This reverts commit 5a5e6d0a17f7c28d8fcb26b943b29aaa0f4d922f.

theis999 commented:
I have added a stack reset on open file, as that left stack in an obviously wrong state. Which i think fixes the first issue you highlighted.
I can reproduce the other error. Correcting it may take a while.

theis999 committed:
Added stack clear when changing files.
Fixed an issue when adding steps through undo/redo as the selected row was not close to the added step.

MortenTobiasNielsen commented on src/cMain.cpp:
```diff
@@ -2278,17 +2280,9 @@ void cMain::UpdateStepGrid(int row, StepParameters* stepParameters)
-	{
-		StepGridData.push_back(*stepParameters);
-	}
+	auto it1 = StepGridData.begin();
+	it1 += row;
+	StepGridData.insert(it1, *stepParameters);
```
Are you sure this works? I would have guessed that I added the branch for a reason.

theis999 replied:
I am certain. All this did was push a row to different point in StepGridData than the grid_steps.

The branching doesn't make sense anyway. Since we already have a row to insert at. 
Overriding that row in a method this far down in the call-stack is incredibly confusing.

MortenTobiasNielsen replied:
Fair point, seems like all the refactoring have made this obsolete and possibly incorrect. Just somewhat strange, that it haven't caused any problems for me, now that I have used to the tool pretty extensively.

MortenTobiasNielsen replied:
Fair point, seems like all the refactoring have made this obsolete and possibly incorrect. Just somewhat strange, that it haven't caused any problems for me, now that I have used to the tool pretty extensively.

theis999 replied:
It wouldn't cause any issues as long as there isn't a discrepancy between the row you interact on and the row that is selected. It is really only Undo/Redo and import-steps that would insert rows invariant of row selection.

MortenTobiasNielsen commented on src/cMain.cpp:
```diff
@@ -2853,9 +2847,9 @@ void cMain::OnSkipClicked(wxCommandEvent& event)
 {
 	grid_steps->ClearSelection();
-	for (auto& [r, _] : rows)
+	for (auto& [row, _] : rows)
```
It is somewhat strange extracting something from the rows list isn't a row, but instead that object has a row property in it. I know it is mostly semantics, but it is somewhat confusing to read, given that it doesn't really make sense.

theis999 replied:
I think i can see what you mean. It is not entirely obvious what each part of the data type represent:
vector is the list of rows
tuple is a concatenation of row index and row data
StepParameters is the row data

`for (auto& [index, data] : rows)` might be more semantically correct.

MortenTobiasNielsen replied:
Agreed, that makes more sense.

MortenTobiasNielsen replied:
Agreed, that makes more sense.

MortenTobiasNielsen commented on src/cMain.cpp:
```diff
@@ -844,7 +846,7 @@ vector< tuple<int, StepParameters>> cMain::DeleteSteps(wxArrayInt steps, bool au
 
 		StepGridData.erase(iStart, iEnd);
 	}
```
Add a new line.

MortenTobiasNielsen reviewed (changes requested):
How does this resolve the "step thinks it can mine a building which haven't been build yet" problem?

theis999 committed:
Added empty line
Changed the local name in SelectRowsInGrid to more readable.

theis999 commented:
It solves by correcting the underlying data structure StepGridData. The issue you found was just a symptom, of the gui not representing the data correctly while the real issue was that the underlying structure was updated incorrectly.

Because you deleted all steps you ended in a situation where no rows were selected in the steps grid. When redo'ing steps these steps would not exclusively be added at the end of the steps list. However as there was no selection, the redone step was added correctly in steps_grid but at the end of StepGridData.

So if you ran a generate steps you'd see that the steps were different from what was in the GUI.
